llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -4745,6 +4745,40 @@ static Instruction *foldICmpAndXX(ICmpInst &I, const SimplifyQuery &Q,
                           Constant::getNullValue(Op1->getType()));
   }
 
+  if (!ICmpInst::isSigned(Pred))
+    return nullptr;
+
+  KnownBits KnownY = IC.computeKnownBits(A, /*Depth=*/0, &I);
+  // (X & NegY) spred X --> (X & NegY) upred X
+  if (KnownY.isNegative())
+    return new ICmpInst(ICmpInst::getUnsignedPredicate(Pred), Op0, Op1);
+
+  if (Pred != ICmpInst::ICMP_SLE && Pred != ICmpInst::ICMP_SGT)
+    return nullptr;
+
+  if (KnownY.isNonNegative()) {
+    // (X & PosY) s<= X --> X s>= 0
+    if (Pred == ICmpInst::ICMP_SLE)
+      return new ICmpInst(ICmpInst::ICMP_SGE, Op1,
+                          Constant::getNullValue(Op1->getType()));
+    // (X & PosY) s> X --> X s< 0
+    if (Pred == ICmpInst::ICMP_SGT)
+      return new ICmpInst(ICmpInst::ICMP_SLT, Op1,
+                          Constant::getNullValue(Op1->getType()));
```

Comment:
```suggestion
    // (X & PosY) s> X --> X s< 0
      return new ICmpInst(Pred == ICmpInst::ICMP_SLE ? ICmpInst::ICMP_SGE : ICmpInst::ICMP_SLT, Op1,
                          Constant::getNullValue(Op1->getType()));
```
or 
```suggestion
    // (X & PosY) s> X --> X s< 0
      return new ICmpInst(ICmpInst::getSwappedPredicate(Pred), Op1,
                          Constant::getNullValue(Op1->getType()));
```

Reply:
It would be better to canonicalize `X s>= 0` -> `X s> -1` here.

Reply:
Personally prefer to leave canonicalization up to the designed folds. Otherwise think you tend to further complicate the code/dup the code/make it harder to change these things.

Reply:
It's good to directly emit canonical IR when convenient, but I agree that we shouldn't go out of our way when it would require extra code.